src/lib/markdowndb.ts

```diff
@@ -54,170 +51,30 @@ export class MarkdownDB {
-        }
-        fileTagsToInsert.push({ file: id, tag });
-      });
-
-      filesToInsert.push({
-        _id: id,
-        file_path: filePath,
-        extension,
-        url_path: urlPath,
-        filetype,
-        metadata,
-      });
-    }
-
-    Object.entries(filesLinksMap).forEach(([fileId, { url, links }]) => {
-      links.forEach(({ linkSrc, linkType }) => {
-        const destPath = resolveLinkToUrlPath(linkSrc, url);
-        const destFile = filesToInsert.find(
-          (file) => file.url_path === destPath
-        );
-        if (!destFile) {
-          return;
-        }
-        const linkToInsert = {
-          // _id: id,
-          from: fileId,
-          to: destFile._id,
-          link_type: linkType,
-        };
-        linksToInsert.push(linkToInsert);
-      });
-    });
-
-    if (filesToInsert.length >= 500) {
-      for (let i = 0; i < filesToInsert.length; i += 500) {
-        await MddbFile.batchInsert(this.db, filesToInsert.slice(i, i + 500));
-      }
-    } else {
-      await MddbFile.batchInsert(this.db, filesToInsert);
-    }
-
-    // TODO  what happens if some of the files were not inserted?
-    // I guess inserting tags or links with such files used as foreign keys will fail too,
-    // but need to check
-
-    if (tagsToInsert.length >= 500) {
-      for (let i = 0; i < tagsToInsert.length; i += 500) {
-        await MddbTag.batchInsert(this.db, tagsToInsert.slice(i, i + 500));
-      }
-    } else {
-      await MddbTag.batchInsert(this.db, tagsToInsert);
-    }
-
-    if (fileTagsToInsert.length >= 500) {
-      for (let i = 0; i < fileTagsToInsert.length; i += 500) {
-        await MddbFileTag.batchInsert(
-          this.db,
-          fileTagsToInsert.slice(i, i + 500)
-        );
-      }
-    } else {
-      await MddbFileTag.batchInsert(this.db, fileTagsToInsert);
-    }
-
-    if (linksToInsert.length >= 500) {
-      for (let i = 0; i < linksToInsert.length; i += 500) {
-        await MddbLink.batchInsert(this.db, linksToInsert.slice(i, i + 500));
-      }
-    } else {
-      await MddbLink.batchInsert(this.db, linksToInsert);
-    }
+    await resetDatabaseTables(this.db);
+
+    const fileObjects = indexFolder(folderPath, ignorePatterns);
+    const filesToInsert = fileObjects.map(mapFileToInsert);
+    const uniqueTags = getUniqueValues(
+      fileObjects.flatMap((file) => file.tags)
+    );
+    const tagsToInsert = uniqueTags.map((tag) => ({ name: tag }));
+    const linksToInsert = fileObjects
+      .flatMap((fileObject) => {
+        return mapLinksToInsert(filesToInsert, fileObject);
+      })
+      .filter(isLinkToDefined);
+    const fileTagsToInsert = fileObjects.flatMap(mapFileTagsToInsert);
+
+    await MddbFile.batchInsert(this.db, filesToInsert);
+    // @ts-ignore
+    await MddbTag.batchInsert(this.db, tagsToInsert);
+    await MddbFileTag.batchInsert(this.db, fileTagsToInsert);
+    await MddbLink.batchInsert(this.db, getUniqueValues(linksToInsert));
```

Comment:
Can you use getUniqueValues earlier? Like you did for tags. Or the other way round, just be consistent.